File: crates/spikard-http/src/cors.rs
//! CORS (Cross-Origin Resource Sharing) handling
//!
//! Handles CORS preflight requests and adds CORS headers to responses

use crate::CorsConfig;
use axum::body::Body;
use axum::http::{HeaderMap, HeaderValue, Response, StatusCode};
use axum::response::IntoResponse;

/// Check if an origin is allowed by the CORS configuration
///
/// Supports exact matches and wildcard ("*") for any origin
fn is_origin_allowed(origin: &str, allowed_origins: &[String]) -> bool {
    if origin.is_empty() {
        return false;
    }

    allowed_origins
        .iter()
        .any(|allowed| allowed == "*" || allowed == origin)
}

/// Check if a method is allowed by the CORS configuration
///
/// Supports exact matches and wildcard ("*") for any method
fn is_method_allowed(method: &str, allowed_methods: &[String]) -> bool {
    allowed_methods
        .iter()
        .any(|allowed| allowed == "*" || allowed.eq_ignore_ascii_case(method))
}

/// Check if all requested headers are allowed
///
/// Headers are case-insensitive. Supports wildcard ("*") for any header.
fn are_headers_allowed(requested: &[&str], allowed: &[String]) -> bool {
    // If wildcard is in allowed list, all headers are permitted
    if allowed.iter().any(|h| h == "*") {
        return true;
    }

    // Check each requested header is in the allowed list (case-insensitive)
    requested.iter().all(|req_header| {
        allowed
            .iter()
            .any(|allowed_header| allowed_header.eq_ignore_ascii_case(req_header))
    })
}

/// Handle CORS preflight (OPTIONS) request
///
/// Validates the request against the CORS configuration and returns appropriate
/// response or error. Returns 204 No Content on success, 403 Forbidden on validation failure.
pub fn handle_preflight(headers: &HeaderMap, cors_config: &CorsConfig) -> Result<Response<Body>, Box<Response<Body>>> {
    // Get the Origin header
    let origin = headers.get("origin").and_then(|v| v.to_str().ok()).unwrap_or("");

    // Validate origin
    if origin.is_empty() || !is_origin_allowed(origin, &cors_config.allowed_origins) {
        return Err(Box::new(
            (
                StatusCode::FORBIDDEN,
                axum::Json(serde_json::json!({
                    "detail": format!("CORS request from origin '{}' not allowed", origin)
                })),
            )
                .into_response(),
        ));
    }

    // Get requested method
    let requested_method = headers
        .get("access-control-request-method")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");

    // Validate method
    if !requested_method.is_empty() && !is_method_allowed(requested_method, &cors_config.allowed_methods) {
        return Err(Box::new((StatusCode::FORBIDDEN).into_response()));
    }

    // Get requested headers
    let requested_headers_str = headers
        .get("access-control-request-headers")
        .and_then(|v| v.to_str().ok());

    if let Some(req_headers) = requested_headers_str {
        let requested_headers: Vec<&str> = req_headers.split(',').map(|h| h.trim()).collect();

        if !are_headers_allowed(&requested_headers, &cors_config.allowed_headers) {
            return Err(Box::new((StatusCode::FORBIDDEN).into_response()));
        }
    }

    // Build preflight response
    let mut response = Response::builder().status(StatusCode::NO_CONTENT);

    // Add CORS headers
    let headers_mut = response.headers_mut().unwrap();

    // Access-Control-Allow-Origin
    headers_mut.insert(
        "access-control-allow-origin",
        HeaderValue::from_str(origin).unwrap_or_else(|_| HeaderValue::from_static("*")),
    );

    // Access-Control-Allow-Methods
    let methods = cors_config.allowed_methods.join(", ");
    headers_mut.insert(
        "access-control-allow-methods",
        HeaderValue::from_str(&methods).unwrap_or_else(|_| HeaderValue::from_static("*")),
    );

    // Access-Control-Allow-Headers
    let allowed_headers = cors_config.allowed_headers.join(", ");
    headers_mut.insert(
        "access-control-allow-headers",
        HeaderValue::from_str(&allowed_headers).unwrap_or_else(|_| HeaderValue::from_static("*")),
    );

    // Access-Control-Max-Age (optional)
    if let Some(max_age) = cors_config.max_age {
        headers_mut.insert(
            "access-control-max-age",
            HeaderValue::from_str(&max_age.to_string()).unwrap(),
        );
    }

    // Access-Control-Allow-Credentials (optional)
    if let Some(true) = cors_config.allow_credentials {
        headers_mut.insert("access-control-allow-credentials", HeaderValue::from_static("true"));
    }

    Ok(response.body(Body::empty()).unwrap())
}

/// Add CORS headers to a successful response
///
/// Adds Access-Control-Allow-Origin, Access-Control-Expose-Headers, and
/// Access-Control-Allow-Credentials based on the configuration
pub fn add_cors_headers(response: &mut Response<Body>, origin: &str, cors_config: &CorsConfig) {
    let headers = response.headers_mut();

    // Access-Control-Allow-Origin
    if let Ok(origin_value) = HeaderValue::from_str(origin) {
        headers.insert("access-control-allow-origin", origin_value);
    }

    // Access-Control-Expose-Headers (optional)
    if let Some(ref expose_headers) = cors_config.expose_headers {
        let expose = expose_headers.join(", ");
        if let Ok(expose_value) = HeaderValue::from_str(&expose) {
            headers.insert("access-control-expose-headers", expose_value);
        }
    }

    // Access-Control-Allow-Credentials (optional)
    if let Some(true) = cors_config.allow_credentials {
        headers.insert("access-control-allow-credentials", HeaderValue::from_static("true"));
    }
}

/// Validate a non-preflight CORS request
///
/// Checks if the Origin header is present and allowed. Returns an error response
/// if validation fails.
pub fn validate_cors_request(headers: &HeaderMap, cors_config: &CorsConfig) -> Result<(), Box<Response<Body>>> {
    let origin = headers.get("origin").and_then(|v| v.to_str().ok()).unwrap_or("");

    if !origin.is_empty() && !is_origin_allowed(origin, &cors_config.allowed_origins) {
        return Err(Box::new(
            (
                StatusCode::FORBIDDEN,
                axum::Json(serde_json::json!({
                    "detail": format!("CORS request from origin '{}' not allowed", origin)
                })),
            )
                .into_response(),
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_cors_config() -> CorsConfig {
        CorsConfig {
            allowed_origins: vec!["https://example.com".to_string()],
            allowed_methods: vec!["GET".to_string(), "POST".to_string()],
            allowed_headers: vec!["content-type".to_string(), "authorization".to_string()],
            expose_headers: Some(vec!["x-custom-header".to_string()]),
            max_age: Some(3600),
            allow_credentials: Some(true),
        }
    }

    #[test]
    fn test_is_origin_allowed_exact_match() {
        let allowed = vec!["https://example.com".to_string()];
        assert!(is_origin_allowed("https://example.com", &allowed));
        assert!(!is_origin_allowed("https://evil.com", &allowed));
    }

    #[test]
    fn test_is_origin_allowed_wildcard() {
        let allowed = vec!["*".to_string()];
        assert!(is_origin_allowed("https://example.com", &allowed));
        assert!(is_origin_allowed("https://any-domain.com", &allowed));
    }

    #[test]
    fn test_is_origin_allowed_empty_origin() {
        let allowed = vec!["*".to_string()];
        assert!(!is_origin_allowed("", &allowed));
    }

    #[test]
    fn test_is_method_allowed_case_insensitive() {
        let allowed = vec!["GET".to_string(), "POST".to_string()];
        assert!(is_method_allowed("GET", &allowed));
        assert!(is_method_allowed("get", &allowed));
        assert!(is_method_allowed("POST", &allowed));
        assert!(is_method_allowed("post", &allowed));
        assert!(!is_method_allowed("DELETE", &allowed));
    }

    #[test]
    fn test_is_method_allowed_wildcard() {
        let allowed = vec!["*".to_string()];
        assert!(is_method_allowed("GET", &allowed));
        assert!(is_method_allowed("DELETE", &allowed));
        assert!(is_method_allowed("PATCH", &allowed));
    }

    #[test]
    fn test_are_headers_allowed_case_insensitive() {
        let allowed = vec!["Content-Type".to_string(), "Authorization".to_string()];
        assert!(are_headers_allowed(&["content-type"], &allowed));
        assert!(are_headers_allowed(&["AUTHORIZATION"], &allowed));
        assert!(are_headers_allowed(&["content-type", "authorization"], &allowed));
        assert!(!are_headers_allowed(&["x-custom"], &allowed));
    }

    #[test]
    fn test_are_headers_allowed_wildcard() {
        let allowed = vec!["*".to_string()];
        assert!(are_headers_allowed(&["any-header"], &allowed));
        assert!(are_headers_allowed(&["multiple", "headers"], &allowed));
    }

    #[test]
    fn test_handle_preflight_success() {
        let config = make_cors_config();
        let mut headers = HeaderMap::new();
        headers.insert("origin", HeaderValue::from_static("https://example.com"));
        headers.insert("access-control-request-method", HeaderValue::from_static("POST"));
        headers.insert(
            "access-control-request-headers",
            HeaderValue::from_static("content-type"),
        );

        let result = handle_preflight(&headers, &config);
        assert!(result.is_ok());

        let response = result.unwrap();
        assert_eq!(response.status(), StatusCode::NO_CONTENT);

        let resp_headers = response.headers();
        assert_eq!(
            resp_headers.get("access-control-allow-origin").unwrap(),
            "https://example.com"
        );
        assert!(
            resp_headers
                .get("access-control-allow-methods")
                .unwrap()
                .to_str()
                .unwrap()
                .contains("POST")
        );
        assert_eq!(resp_headers.get("access-control-max-age").unwrap(), "3600");
        assert_eq!(resp_headers.get("access-control-allow-credentials").unwrap(), "true");
    }

    #[test]
    fn test_handle_preflight_origin_not_allowed() {
        let config = make_cors_config();
        let mut headers = HeaderMap::new();
        headers.insert("origin", HeaderValue::from_static("https://evil.com"));
        headers.insert("access-control-request-method", HeaderValue::from_static("GET"));

        let result = handle_preflight(&headers, &config);
        assert!(result.is_err());

        let response = *result.unwrap_err();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
    }

    #[test]
    fn test_handle_preflight_method_not_allowed() {
        let config = make_cors_config();
        let mut headers = HeaderMap::new();
        headers.insert("origin", HeaderValue::from_static("https://example.com"));
        headers.insert("access-control-request-method", HeaderValue::from_static("DELETE"));

        let result = handle_preflight(&headers, &config);
        assert!(result.is_err());

        let response = *result.unwrap_err();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
    }

    #[test]
    fn test_handle_preflight_header_not_allowed() {
        let config = make_cors_config();
        let mut headers = HeaderMap::new();
        headers.insert("origin", HeaderValue::from_static("https://example.com"));
        headers.insert("access-control-request-method", HeaderValue::from_static("POST"));
        headers.insert(
            "access-control-request-headers",
            HeaderValue::from_static("x-forbidden-header"),
        );

        let result = handle_preflight(&headers, &config);
        assert!(result.is_err());

        let response = *result.unwrap_err();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
    }

    #[test]
    fn test_handle_preflight_empty_origin() {
        let config = make_cors_config();
        let headers = HeaderMap::new(); // No origin header

        let result = handle_preflight(&headers, &config);
        assert!(result.is_err());

        let response = *result.unwrap_err();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
    }

    #[test]
    fn test_add_cors_headers() {
        let config = make_cors_config();
        let mut response = Response::new(Body::empty());

        add_cors_headers(&mut response, "https://example.com", &config);

        let headers = response.headers();
        assert_eq!(
            headers.get("access-control-allow-origin").unwrap(),
            "https://example.com"
        );
        assert_eq!(headers.get("access-control-expose-headers").unwrap(), "x-custom-header");
        assert_eq!(headers.get("access-control-allow-credentials").unwrap(), "true");
    }

    #[test]
    fn test_validate_cors_request_allowed() {
        let config = make_cors_config();
        let mut headers = HeaderMap::new();
        headers.insert("origin", HeaderValue::from_static("https://example.com"));

        let result = validate_cors_request(&headers, &config);
        assert!(result.is_ok());
    }

    #[test]
    fn test_validate_cors_request_not_allowed() {
        let config = make_cors_config();
        let mut headers = HeaderMap::new();
        headers.insert("origin", HeaderValue::from_static("https://evil.com"));

        let result = validate_cors_request(&headers, &config);
        assert!(result.is_err());

        let response = *result.unwrap_err();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
    }

    #[test]
    fn test_validate_cors_request_no_origin() {
        let config = make_cors_config();
        let headers = HeaderMap::new(); // No origin header

        // No origin header is allowed (not a CORS request)
        let result = validate_cors_request(&headers, &config);
        assert!(result.is_ok());
    }
}

File: packages/elixir/lib/spikard/cors.ex
defmodule Spikard.Cors do
  @moduledoc """
  CORS (Cross-Origin Resource Sharing) configuration for Spikard.

  This module provides a struct and validation functions for configuring
  CORS behavior in Spikard applications.

  ## Configuration Options

  - `:allowed_origins` - List of allowed origins (required). Use `["*"]` to allow all origins.
  - `:allowed_methods` - List of allowed HTTP methods (required). Use `["*"]` to allow all methods.
  - `:allowed_headers` - List of allowed request headers (optional). Use `["*"]` to allow all headers.
  - `:expose_headers` - List of headers to expose to the client (optional).
  - `:max_age` - Preflight cache duration in seconds (optional).
  - `:allow_credentials` - Whether to include credentials in CORS requests (optional).

  ## Examples

      # Allow any origin with GET and POST
      cors: %{
        allowed_origins: ["*"],
        allowed_methods: ["GET", "POST"]
      }

      # Restrict to specific origins with custom headers
      cors: %{
        allowed_origins: ["https://example.com", "https://app.example.com"],
        allowed_methods: ["GET", "POST", "PUT"],
        allowed_headers: ["content-type", "authorization"],
        expose_headers: ["x-total-count", "x-page"],
        max_age: 3600,
        allow_credentials: true
      }

  ## CORS Preflight

  CORS preflight requests (OPTIONS requests) are automatically handled by the
  Spikard framework. Responses include appropriate headers:

  - `Access-Control-Allow-Origin` - Allowed origin
  - `Access-Control-Allow-Methods` - Allowed methods
  - `Access-Control-Allow-Headers` - Allowed headers
  - `Access-Control-Max-Age` - Preflight cache duration
  - `Access-Control-Allow-Credentials` - Whether credentials are allowed

  Regular (non-OPTIONS) requests also include CORS headers when applicable.
  """

  @typedoc "CORS configuration map"
  @type t :: map()

  @doc """
  Validates a CORS configuration map.

  Returns `{:ok, config}` if the configuration is valid, or
  `{:error, reason}` if there are validation errors.

  Required keys:
  - `:allowed_origins` - Non-empty list of strings
  - `:allowed_methods` - Non-empty list of strings

  Optional keys:
  - `:allowed_headers` - List of strings (default: [])
  - `:expose_headers` - List of strings or nil
  - `:max_age` - Positive integer or nil
  - `:allow_credentials` - Boolean or nil
  """
  @spec validate(map()) :: {:ok, t()} | {:error, String.t()}
  def validate(config) when is_map(config) do
    with :ok <- validate_allowed_origins(config),
         :ok <- validate_allowed_methods(config),
         :ok <- validate_optional_fields(config) do
      {:ok, config}
    end
  end

  def validate(_), do: {:error, "CORS configuration must be a map"}

  @doc """
  Validates a CORS configuration, raising an error if invalid.

  See `validate/1` for detailed validation rules.
  """
  @spec validate!(map()) :: t()
  def validate!(config) do
    case validate(config) do
      {:ok, validated} -> validated
      {:error, reason} -> raise ArgumentError, reason
    end
  end

  # Private helper functions

  defp validate_allowed_origins(config) do
    case Map.get(config, :allowed_origins) do
      nil ->
        {:error, "CORS configuration requires :allowed_origins"}

      origins when is_list(origins) ->
        if Enum.empty?(origins) do
          {:error, "CORS :allowed_origins cannot be empty"}
        else
          case Enum.all?(origins, &is_binary/1) do
            true -> :ok
            false -> {:error, "CORS :allowed_origins must be a list of strings"}
          end
        end

      _ ->
        {:error, "CORS :allowed_origins must be a list"}
    end
  end

  defp validate_allowed_methods(config) do
    case Map.get(config, :allowed_methods) do
      nil ->
        {:error, "CORS configuration requires :allowed_methods"}

      methods when is_list(methods) ->
        if Enum.empty?(methods) do
          {:error, "CORS :allowed_methods cannot be empty"}
        else
          case Enum.all?(methods, &is_binary/1) do
            true -> :ok
            false -> {:error, "CORS :allowed_methods must be a list of strings"}
          end
        end

      _ ->
        {:error, "CORS :allowed_methods must be a list"}
    end
  end

  defp validate_optional_fields(config) do
    with :ok <- validate_allowed_headers(config),
         :ok <- validate_expose_headers(config),
         :ok <- validate_max_age(config),
         :ok <- validate_allow_credentials(config) do
      :ok
    end
  end

  defp validate_allowed_headers(config) do
    case Map.get(config, :allowed_headers) do
      nil ->
        :ok

      headers when is_list(headers) ->
        if Enum.all?(headers, &is_binary/1) do
          :ok
        else
          {:error, "CORS :allowed_headers must be a list of strings"}
        end

      _ ->
        {:error, "CORS :allowed_headers must be a list or nil"}
    end
  end

  defp validate_expose_headers(config) do
    case Map.get(config, :expose_headers) do
      nil ->
        :ok

      headers when is_list(headers) ->
        if Enum.all?(headers, &is_binary/1) do
          :ok
        else
          {:error, "CORS :expose_headers must be a list of strings"}
        end

      _ ->
        {:error, "CORS :expose_headers must be a list or nil"}
    end
  end

  defp validate_max_age(config) do
    case Map.get(config, :max_age) do
      nil -> :ok
      age when is_integer(age) and age > 0 -> :ok
      _ -> {:error, "CORS :max_age must be a positive integer or nil"}
    end
  end

  defp validate_allow_credentials(config) do
    case Map.get(config, :allow_credentials) do
      nil -> :ok
      true -> :ok
      false -> :ok
      _ -> {:error, "CORS :allow_credentials must be a boolean or nil"}
    end
  end
end
